lib/db/backend/leveldb_backend.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+const (
+	dbFlushBatch = 4 << MiB
+)
+
+// leveldbBackend implements Backend on top of a leveldb
+type leveldbBackend struct {
+	ldb     *leveldb.DB
+	closeWG sync.WaitGroup
+}
+
+func (b *leveldbBackend) NewReadTransaction() (ReadTransaction, error) {
+	snap, err := b.ldb.GetSnapshot()
+	if err != nil {
+		return nil, wrapLeveldbErr(err)
+	}
+	return leveldbSnapshot{
+		snap: snap,
+		rel:  newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *leveldbBackend) NewWriteTransaction() (WriteTransaction, error) {
+	snap, err := b.ldb.GetSnapshot()
+	if err != nil {
+		return nil, wrapLeveldbErr(err)
+	}
+	return &leveldbTransaction{
+		leveldbSnapshot: leveldbSnapshot{
+			snap: snap,
+			rel:  newReleaser(&b.closeWG),
+		},
+		ldb: b.ldb,
+		rel: newReleaser(&b.closeWG),
+	}, nil
+}
+
+func (b *leveldbBackend) Close() error {
+	b.closeWG.Wait()
+	return wrapLeveldbErr(b.ldb.Close())
+}
+
+func (b *leveldbBackend) Get(key []byte) ([]byte, error) {
+	val, err := b.ldb.Get(key, nil)
+	return val, wrapLeveldbErr(err)
+}
+
+func (b *leveldbBackend) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	return b.ldb.NewIterator(util.BytesPrefix(prefix), nil), nil
+}
+
+func (b *leveldbBackend) NewRangeIterator(first, last []byte) (Iterator, error) {
+	return b.ldb.NewIterator(&util.Range{Start: first, Limit: last}, nil), nil
+}
+
+func (b *leveldbBackend) Put(key, val []byte) error {
+	return wrapLeveldbErr(b.ldb.Put(key, val, nil))
+}
+
+func (b *leveldbBackend) Delete(key []byte) error {
+	return wrapLeveldbErr(b.ldb.Delete(key, nil))
+}
+
+// leveldbSnapshot implements backend.ReadTransaction
+type leveldbSnapshot struct {
+	snap *leveldb.Snapshot
+	rel  *releaser
+}
+
+func (l leveldbSnapshot) Get(key []byte) ([]byte, error) {
+	val, err := l.snap.Get(key, nil)
+	return val, wrapLeveldbErr(err)
+}
+
+func (l leveldbSnapshot) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	return l.snap.NewIterator(util.BytesPrefix(prefix), nil), nil
+}
+
+func (l leveldbSnapshot) NewRangeIterator(first, last []byte) (Iterator, error) {
+	return l.snap.NewIterator(&util.Range{Start: first, Limit: last}, nil), nil
+}
+
+func (l leveldbSnapshot) Release() {
+	defer l.rel.Release()
+	l.snap.Release()
+}
+
+// leveldbTransaction implements backend.WriteTransaction using a batch (not
+// an actual leveldb transaction)
+type leveldbTransaction struct {
+	leveldbSnapshot
+	ldb   *leveldb.DB
+	batch *leveldb.Batch
+	rel   *releaser
+}
+
+func (t *leveldbTransaction) Delete(key []byte) error {
+	if t.batch == nil {
+		t.batch = new(leveldb.Batch)
+	}
+	t.batch.Delete(key)
+	return wrapLeveldbErr(t.checkFlush())
+}
+
+func (t *leveldbTransaction) Put(key, val []byte) error {
+	if t.batch == nil {
+		t.batch = new(leveldb.Batch)
+	}
+	t.batch.Put(key, val)
+	return wrapLeveldbErr(t.checkFlush())
+}
+
+func (t *leveldbTransaction) Commit() error {
+	defer t.rel.Release()
+	defer t.leveldbSnapshot.Release()
```

Comment:
This will result in a synchronous i/o most likely so I think inlining of tiny functions is a teensy consideration here. But sure it could be rewritten with a temp variable.